core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -102,8 +99,15 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   // hadoop token manager used by some sub-classes (e.g. Mesos)
   def hadoopDelegationTokenManager: Option[HadoopDelegationTokenManager] = None
 
-  // Hadoop delegation tokens to be sent to the executors.
-  val hadoopDelegationCreds: Option[Array[Byte]] = getHadoopDelegationCreds()
+  def hadoopCredentialRenewer: Option[HadoopCredentialRenewer] = None
+
+  // Hadoop delegation tokens to be sent to the executors, can be updated as necessary.
+  var renewableDelegationTokens: Option[RenewableDelegationTokens] =
```

Comment:
This is another part where the interface is brittle and a bit confusing. This should really be a private field, because there should be a single way for sub-classes to update it (send a `UpdateDelegationTokens` message). But it's not because Mesos needs to poke at it because the rest of the API is all a bit weird.

If instead you follow my original suggestion (have a abstract `initializeDelegationTokens()` method that Mesos overrides), then everything becomes a lot more localized and cleaner.